src/motor-operations.cc

```diff
@@ -85,11 +86,40 @@ static char test_acceleration_ok(float acceleration) {
 }
 #endif
 
+typedef struct HistoryPositionInfo {
+  HistoryPositionInfo () : position_steps(0), sign(1) {}
+  int position_steps;
+  uint32_t fraction;
+  signed char sign;
+} HistoryPositionInfo;
+
+// Store the required informations needed to backtrack the absolute position.
+struct MotionQueueMotorOperations::HistorySegment {
+  HistoryPositionInfo pos_info[MOTION_MOTOR_COUNT];
+  unsigned short aux_bits;
+};
+
+MotionQueueMotorOperations::
+  MotionQueueMotorOperations(MotionQueue *backend)
+  : backend_(backend),
+    shadow_queue_(new std::deque<struct HistorySegment>()) {
+  // Initialize the history queue.
+  struct HistorySegment new_hs = {};
+  shadow_queue_->push_front(new_hs);
+}
+
+MotionQueueMotorOperations::~MotionQueueMotorOperations() {
+  delete shadow_queue_;
+}
+
 void MotionQueueMotorOperations::EnqueueInternal(const LinearSegmentSteps &param,
                                                  int defining_axis_steps) {
   struct MotionSegment new_element = {};
   new_element.direction_bits = 0;
 
+  // Get the last history segment pushed
```

Comment:
In comments: Don't describe what the code is already saying ('get last history segment'), but rather why that is. Here, because the final position is relative to the previous position. So something like
   // The new segment is based on the previous position
or similar.